# ЛР-1. Работа с данными: очистка и отбор признаков

Датасет: ID_data_mass_18122012 (скважинные измерения).  
Целевые переменные: `G_total`, `КГФ`.

In [1]:
from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from mlcore.tabular.workflow import run_tabular_feature_review
from mlcore.tabular.preprocessing import normalize_missing_values
from mlcore.utils.notebook import lab_paths

pd.set_option('display.max_columns', 40)
pd.set_option('display.width', 200)

In [2]:
ROOT, LAB_DIR, ASSETS_DIR = lab_paths(1)
DATA_PATH = LAB_DIR / 'data/dataset.csv'

## 1. Загрузка и первичный осмотр

In [3]:
df = pd.read_csv(DATA_PATH, sep=',', decimal=',', header=1, skiprows=[2], na_values=['-', ''])
print(f'Shape: {df.shape}')
df.head()

Shape: (185, 34)


,Unnamed: 0,Unnamed: 1,Глубина манометра,Dшт,Руст,Рзаб,Pлин,Руст.1,Рзаб.1,Рлин,Туст,Тна шлейфе,Тзаб,Tлин,Дебит газа,Дебит ст. конд.,Дебит воды,Дебит смеси,Дебит гааз,Дебит кон нестабильный,Дебит воды.1,Нэф,Рпл. Тек (послед точка на КВД),Рпл. Тек (Расчет по КВД),Рпл. Тек (Карноухов),Pсб,Pсб.1,Ro_g,Ro_c,Ro_w,Удельная плотность газа,G_total,КГФ,КГФ.1
0,804,2008-06-05 00:00:00,3576.3,7.94,249.60000000000002,370.1,101.8,249,359.6,101.8,53,31.7,103.2,32.5,214.704,83.6,0.4,231.24884,1610.368866,131.3,0.4,56.8,45.25,56.5,NaN,93.6,92.3760177646188,0.8060166657971012,801,1000.0,0.6694490579710143,2.782622710802093,311.9094,NaN
1,804,2008-06-06 00:00:00,3576.3,9.53,233.5,364.6,101.3,231,338.1,102.4,58.8,37.6,103,38.6,290.594,104.7,1.4,309.00166,2310.226429,158.5,1.4,56.8,45.25,56.5,NaN,92.9,91.68517147791759,0.8060166657971012,801,1000.0,0.6694490579710143,3.6977813307944776,288.6003,NaN
2,804,2008-06-07 00:00:00,3576.3,11.11,213.4,357.1,101.6,211,314.8,100.6,63.6,42.8,102.6,43.4,368.042,114.3,1.9,388.11498000000006,3039.48968,172.3,1.9,56.8,45.25,56.5,NaN,91.4,90.20478657784358,0.8060166657971012,801,1000.0,0.6694490579710143,4.515072751311305,248.7906,NaN
3,804,2008-06-08 00:00:00,3576.3,12.7,191.6,347.4,98.4,187,291.5,99,64.7,46.2,102,46.1,434.663,121.3,3.3,455.21451999999994,3824.082686,181.5,3.3,56.8,45.25,56.5,NaN,89.2,88.03355539106835,0.8060166657971012,801,1000.0,0.6694490579710143,5.217672708395432,223.5591,NaN
4,804,2008-06-09 00:00:00,3576.3,14.29,171.9,337.69999999999993,99.2,169,270.3,99.5,64,49.4,104.4,49.9,483.283,129.8,4.6,504.58777999999995,4299.104676,190.6,4.7,56.8,45.25,56.5,NaN,89.7,88.52701702442636,0.8060166657971012,801,1000.0,0.6694490579710143,5.765092040467829,215.1486,NaN


## 2. Очистка данных

- Удаляем столбцы-метаданные (`Unnamed: 0` — номер скважины, `Unnamed: 1` — дата, `КГФ.1` — дубликат).
- Конвертируем строковые столбцы с десятичной запятой в числовые.

In [4]:
df = df.drop(columns=['Unnamed: 0', 'Unnamed: 1', 'КГФ.1'])

# Строки с десятичной запятой -> числа
for col in df.select_dtypes(include=['object', 'string']).columns:
    df[col] = df[col].astype(str).str.replace(',', '.', regex=False)
    df[col] = pd.to_numeric(df[col], errors='coerce')

print(f'Shape after cleanup: {df.shape}')
print(f'dtypes: all numeric = {(df.dtypes != "object").all()}')

Shape after cleanup: (185, 31)
dtypes: all numeric = True


In [5]:
# Пропуски
df_norm = normalize_missing_values(df)
missing = df_norm.isna().sum().sort_values(ascending=False)
missing_pct = (missing / len(df_norm) * 100).round(1)
pd.DataFrame({'na_count': missing, 'na_pct': missing_pct}).query('na_count > 0')

,na_count,na_pct
КГФ,162,87.6
G_total,162,87.6
Ro_w,162,87.6
Ro_g,162,87.6
Удельная плотность газа,92,49.7
Ro_c,92,49.7
Тна шлейфе,62,33.5
Рпл. Тек (Карноухов),50,27.0
Рпл. Тек (Расчет по КВД),27,14.6
Рлин,11,5.9


## 3. Полный анализ признаков

Используем `run_tabular_feature_review` — единый вызов, который выполняет:
- Удаление строк где оба таргета пусты
- Gain ratio по каждому таргету
- Корреляционную матрицу (абсолютные значения)
- Описательную статистику и квартили
- Сохранение всех графиков

In [6]:
TARGETS = ['G_total', 'КГФ']

artifacts = run_tabular_feature_review(
    df, target_columns=TARGETS, plots_output_dir=ASSETS_DIR, score_bins=10
)

print(f'Cleaned shape: {artifacts.cleaned_df.shape}')
print(f'Features: {len(artifacts.feature_names)}')

Cleaned shape: (23, 31)
Features: 29


### 3.1. Gain ratio (информативность признаков)

In [7]:
artifacts.score_mean.to_frame('mean_gain_ratio')

,mean_gain_ratio
Дебит смеси,0.755734
Дебит газа,0.729341
Tлин,0.724360
Дебит гааз,0.724360
Dшт,0.716089
Руст,0.692986
Рзаб.1,0.691023
Дебит ст. конд.,0.689752
Тзаб,0.688024
Тна шлейфе,0.682534


### 3.2. Корреляционная матрица

![heatmap](../assets/abs_correlation_heatmap.png)

In [8]:
# Пары с корреляцией > 0.95 (мультиколлинеарность)
corr = artifacts.abs_corr_matrix
high_corr = []
for i in range(len(corr)):
    for j in range(i + 1, len(corr)):
        if corr.iloc[i, j] > 0.95:
            high_corr.append((corr.index[i], corr.columns[j], round(corr.iloc[i, j], 3)))

pd.DataFrame(high_corr, columns=['Feature A', 'Feature B', '|r|']).sort_values('|r|', ascending=False)

,Feature A,Feature B,|r|
7,Дебит газа,Дебит смеси,1.000
14,Pсб,Pсб.1,1.000
15,Ro_g,Удельная плотность газа,1.000
0,Глубина манометра,Нэф,0.999
2,Руст,Руст.1,0.999
11,Дебит смеси,G_total,0.996
13,Рпл. Тек (Карноухов),Ro_c,0.996
8,Дебит газа,G_total,0.993
10,Дебит воды,Дебит воды.1,0.992
5,Pлин,Рлин,0.977


### 3.3. Распределения признаков

Гистограммы с линиями Q1/Q3 сохранены в `assets/distributions/`.

### 3.4. Описательная статистика

In [9]:
artifacts.descriptive_stats

,count,mean,std,min,25%,50%,75%,max
Глубина манометра,22.0,3631.127273,34.389356,3576.300000,3594.725000,3650.000000,3653.000000,3653.000000
Dшт,23.0,11.804348,2.569128,7.940000,9.530000,11.110000,14.290000,15.880000
Руст,23.0,242.507904,59.287573,138.866667,195.550000,249.600000,291.663710,337.108108
Рзаб,22.0,367.727273,67.399316,239.300000,316.025000,379.400000,422.075000,458.600000
Pлин,23.0,98.861344,12.094472,73.306638,94.416770,101.300000,107.309079,121.682930
Руст.1,23.0,240.673913,59.242876,136.000000,194.500000,249.000000,293.000000,333.000000
Рзаб.1,22.0,359.990909,70.700824,239.300000,292.825000,374.100000,422.050000,458.500000
Рлин,23.0,98.286720,12.396412,72.579258,90.945790,100.600000,106.480000,123.616800
Туст,23.0,60.808696,10.192908,37.000000,52.500000,63.600000,68.500000,75.000000
Тна шлейфе,18.0,53.494444,31.771713,14.000000,34.250000,38.400000,80.075000,110.200000


## 4. Отбор признаков

Критерии:
- **Gain ratio**: средний gain ratio > 0 (информативность для таргетов)
- **Корреляция**: если два признака коррелируют > 0.95, оставляем более информативный
- **Пропуски**: > 80% пропусков — удаляем
- **Вариативность**: константные или почти константные — удаляем

In [10]:
scores = artifacts.score_mean
missing_rate = artifacts.cleaned_df[artifacts.feature_names].isna().mean()

# Признаки для удаления
DROP = []
KEEP = []
reasons = {}

for feat in artifacts.feature_names:
    if missing_rate[feat] > 0.8:
        DROP.append(feat)
        reasons[feat] = f'>{80}% пропусков ({missing_rate[feat]:.0%})'
    elif scores.get(feat, 0) == 0.0:
        DROP.append(feat)
        reasons[feat] = 'gain ratio = 0 (не информативен)'
    else:
        KEEP.append(feat)
        reasons[feat] = f'gain ratio = {scores.get(feat, 0):.3f}'

# Проверяем мультиколлинеарность среди KEEP
to_remove = set()
for a, b, r in high_corr:
    if a in KEEP and b in KEEP:
        # Удаляем менее информативный
        worse = b if scores.get(a, 0) >= scores.get(b, 0) else a
        if worse not in to_remove:
            to_remove.add(worse)
            reasons[worse] = f'мультиколлинеарность (|r|={r:.3f})'

for feat in to_remove:
    KEEP.remove(feat)
    DROP.append(feat)

print(f'KEEP: {len(KEEP)} признаков')
print(f'DROP: {len(DROP)} признаков')
print()

decision = pd.DataFrame([
    {'Признак': f, 'Решение': 'Оставить' if f in KEEP else 'Удалить', 'Обоснование': reasons[f]}
    for f in artifacts.feature_names
])
decision

KEEP: 18 признаков
DROP: 11 признаков



,Признак,Решение,Обоснование
0,Глубина манометра,Оставить,gain ratio = 0.507
1,Dшт,Оставить,gain ratio = 0.716
2,Руст,Оставить,gain ratio = 0.693
3,Рзаб,Удалить,мультиколлинеарность (|r|=0.951)
4,Pлин,Удалить,мультиколлинеарность (|r|=0.977)
5,Руст.1,Удалить,мультиколлинеарность (|r|=0.999)
6,Рзаб.1,Удалить,мультиколлинеарность (|r|=0.973)
7,Рлин,Оставить,gain ratio = 0.653
8,Туст,Оставить,gain ratio = 0.661
9,Тна шлейфе,Оставить,gain ratio = 0.683


## 5. Выводы

1. Датасет содержит 185 наблюдений, но только 23 имеют непустые значения обоих таргетов.
2. Gain ratio позволил выделить наиболее информативные признаки для предсказания G_total и КГФ.
3. Обнаружены группы сильно коррелирующих признаков — из каждой группы оставлен наиболее информативный.
4. Итоговый набор признаков готов для построения моделей.